generateExpressionTests.py

```diff
@@ -0,0 +1,291 @@
+            "  %s %s = stan::math::as_array_or_scalar(stan::test::make_arg<%s>())-1;\n"
+            % (arg_type, var_name, arg_type)
+        )
+    else:
+        return "  %s %s = stan::test::make_arg<%s>();\n" % (
+            arg_type,
+            var_name,
+            arg_type,
+        )
+
+
+def save_tests_in_files(N_files, tests):
+    """
+    Saves tests in files
+    :param N_files: number of files to distribute tests into
+    :param tests: list of test sources
+    """
+    for i in range(N_files):
+        start = i * len(tests) // N_files
+        end = (i + 1) * len(tests) // N_files
+        with open(src_folder + "tests%d_test.cpp" % i, "w") as out:
+            out.write("#include <test/expressions/expression_test_helpers.hpp>\n\n")
+            for test in tests[start:end]:
+                out.write(test)
+
+
+def main(functions=(), j=1):
+    """
+    Generates expression tests. Functions that do not support expressions yet are listed
+    in stan_math/tests/expressions/stan_math_sigs_exceptions.expected
+
+    For every signature prim, rev and fwd instantiations are tested (with all scalars
+    of type double/var/fvar<double>). Tests check the following:
+     - signatures can be compiled with expressions arguments
+     - results when given expressions are same as when given plain matrices
+       (including derivatives)
+     - functions evaluate expressions at most once
+
+    :param functions: functions to generate tests for. Default: all
+    :param j: number of files to split tests in
+    """
+    ignored = get_ignored_signatures()
+
+    test_n = {}
+    tests = []
+    for signature in get_signatures():
+        return_type, function_name, function_args = parse_signature(signature)
+        if signature in ignored:
+            continue
+        for arg2test in args2test:
+            if arg2test in function_args:
+                break
+        else:
+            continue
+        if functions and function_name not in functions:
+            continue
+        func_test_n = test_n.get(function_name, 0)
+        test_n[function_name] = func_test_n + 1
+
+        if function_name.endswith("_rng"):
+            function_args.append("rng")
+
+        test_code = ""
+        for overload, scalar in (
+            ("Prim", "double"),
+            ("Rev", "stan::math::var"),
+            ("Fwd", "stan::math::fvar<double>"),
+        ):
+            if function_name.endswith("_rng") and overload != "Prim":
+                continue
+            test_code += "TEST(ExpressionTest%s, %s%d){\n" % (
+                overload,
+                function_name,
+                func_test_n,
+            )
+
+            for n, arg in enumerate(function_args):
+                test_code += make_arg_code(arg, scalar, "arg_mat%d" % n, function_name)
+            test_code += "  auto res_mat = stan::math::%s(" % function_name
+            for n in range(len(function_args) - 1):
+                test_code += "arg_mat%d, " % n
```

Comment:
THere are a lot of bits of code that look like things that could be replaced by ample use of the join function:

```
",".join(...)
```

With the templated code generation stuff I like the high level templates that you fill in:

```
"""
TEST({test_group}, {test_name}) {
  {argument_declarations}
  {expression_declarations}

  {evaluate_to_value}
  {evaluate_to_expression}

  compare(value, expression);
}
""".format(...)
```

Etc so it's easy to read the basic template and then look around to fill things in. But this is optional if you want to go all the way there. I think we should look at using ```",".join(...)``` more though.

Reply:
In python many things can be done in one-liners. That does not mean they should be, as they tend to be harder to read.
For example the loop you commented on could be rewritten as 
```
",".join("arg_mat%d, " % n for n in range(len(function_args) - 1))
```
There are also more extreme examples. I like the "basic template" idea tho.